git-repo-prep/index.ts

```diff
@@ -0,0 +1,214 @@
+}
+
+function getCollectionMeta(colPath: string): ColMeta {
+  const metaXpath = '//*[local-name() = "metadata"]'
+  const titleXpath = `${metaXpath}/md:title`
+  const licenseXpath = `${metaXpath}/md:license`
+
+  const collection = fs.readFileSync(colPath, { encoding: 'utf-8' })
+  const doc = new DOMParser().parseFromString(collection)
+  const title = expectValue(selectOne(titleXpath, doc).textContent, 'No title')
+  const licenseEl = selectOne(licenseXpath, doc)
+
+  return {
+    title: title,
+    license: getLicense(licenseEl)
+  }
+}
+
+function getBookMeta(bookPath: string): BookMeta {
+  const booksXmlPath = getBooksXmlPath(bookPath)
+  const booksXml = fs.readFileSync(booksXmlPath, { encoding: 'utf-8' })
+  const doc = new DOMParser().parseFromString(booksXml)
+  const slugsMeta: SlugMeta[] = []
+  const slugs = select('//*[@slug]', doc) as Element[]
+  for (const slug of slugs) {
+    const slugName = expectValue(slug.getAttribute('slug'), 'Slug not found')
+    const href = expectValue(slug.getAttribute('href'), 'Slug href not found')
+    const collectionId = expectValue(
+      slug.getAttribute('collection-id'),
+      'collection-id not found'
+    )
+    const colPath = getColPath(booksXmlPath, href)
+    const colMeta = getCollectionMeta(colPath)
+    slugsMeta.push({ slugName, collectionId, href, colMeta })
+  }
+  return { slugsMeta }
+}
+
+function getBooksXmlPath(bookPath: string) {
+  return path.join(bookPath, 'META-INF', 'books.xml')
+}
+
+function getColPath(booksXmlPath: string, href: string): string {
+  return path.join(path.dirname(booksXmlPath), href)
+}
+
+function writeReadme(meta: BookMeta, bookPath: string, templateDir: string) {
+  fs.writeFileSync(
+    path.join(bookPath, 'README.md'),
+    meta.slugsMeta.length === 1
+      ? createBookReadme(templateDir, meta.slugsMeta[0])
+      : createBundleReadme(templateDir, meta.slugsMeta)
+  )
+}
+
+function createBookReadme(templateDir: string, slugMeta: SlugMeta) {
+  const templatePath = path.join(templateDir, 'book.md')
+  const replacements = {
+    'book_title': slugMeta.colMeta.title,
+    'book_link': `${BOOK_WEB_ROOT}${encodeURIComponent(slugMeta.slugName)}`,
+    'license_text': slugMeta.colMeta.license.text,
+    'license_type': slugMeta.colMeta.license.type,
+    'license_version': slugMeta.colMeta.license.version
+  }
+  return populateTemplate(templatePath, replacements)
+}
+
+function createBundleReadme(templateDir: string, slugsMeta: SlugMeta[]) {
+  const templatePath = path.join(templateDir, 'bundle.md')
+  const replacements = {
+    'book_titles':
+      slugsMeta.map((s, i) =>
+        i === slugsMeta.length - 1 ? `and ${s.colMeta.title}` : s.colMeta.title
+      ).join(slugsMeta.length > 2 ? ', ' : ' '),
+    'book_links':
+      slugsMeta.map(s =>
+        `- _${s.colMeta.title}_ [online](${BOOK_WEB_ROOT}${encodeURIComponent(s.slugName)})`
+      ).join('\n'),
+    'license_text': slugsMeta[0].colMeta.license.text,
```

Comment:
This seems to assume that all books in a repo have the same license. I'm not sure if that is true. Could this error if the licenses are different?

Maybe the python code handles this? I see a line with the following:

```python
raise Exception('Licenses differ between collections')
```

Reply:
I figured it was better to have the check closer to the assumption, so I moved it into the typescript.